### PR TITLE
Customize dio adapter

### LINKED ISSUE #138: Integrating the http/2 protocol into the plugin
**Description**
Could you add http/2 support to the plugin? I looked at your plugin source code and saw that you are using Dio. on the page of this plugin there is information about using this protocol.

**Basic example**
```
dependencies:
  dio: ^4.0.0
  dio_http2_adapter: ^2.0.0
```
import 'package:dio/dio.dart';
import 'package:dio_http2_adapter/dio_http2_adapter.dart';

```
  var dio = Dio()
    ..options.baseUrl = 'https://google.com'
    ..interceptors.add(LogInterceptor())
    ..httpClientAdapter = Http2Adapter(
      ConnectionManager(
        idleTimeout: 10000,
        // Ignore bad certificate
        onClientCreate: (_, config) => config.onBadCertificate = (_) => true,
      ),
    );
```

### PR DESCRIPTION
**Pull request by ahmednfwela**
Fixes #138

this adds a new `MeiliSearchClient.withCustomDio` which can customize adapter and interceptors

- [ ] Added Tests

### CODE REVIEW

File: test/utils/client.dart [hunk 1]
```diff
@@ -30,11 +31,23 @@ Future<void> deleteAllKeys() async {
   }
 }
 
-Future<void> setUpClient() async {
+Future<void> setUpClient({bool isHttp2 = false}) async {
```
Comment (brunoocasali) [inline test/utils/client.dart:34]: Can you set up some tests that actually use this new implementation?

Comment (brunoocasali) [reply test/utils/client.dart:34]: Also, can you write a small "how to" piece in the README? So the other users will know how to use it?

**Comment by ahmednfwela** [comment]
I added some tests in the health_test.dart but I have no idea how to make the server accept http2, can you help with that ?

File: lib/src/http_request_impl.dart [hunk 2]
```diff
@@ -6,8 +6,13 @@ import 'exception.dart';
 const bool _kIsWeb = bool.fromEnvironment('dart.library.js_util');
 
 class HttpRequestImpl implements HttpRequest {
-  HttpRequestImpl(this.serverUrl, this.apiKey, [this.connectTimeout])
-      : dio = Dio(
+  HttpRequestImpl(
```
Comment (brunoocasali) [inline lib/src/http_request_impl.dart:9]: Just checking, if the user uses the new constructor, all the headers, auth, user agent, and others will still be sent correctly, right?

File: test/health_test.dart [hunk 3]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
+      test('when the url is not valid', () async {
+        expect(client.health(), throwsException);
+      });
 
-    test('when the url is not valid with isHealthy', () async {
-      var health = await client.isHealthy();
-      expect(health, false);
+      test('when the url is not valid with isHealthy', () async {
+        var health = await client.isHealthy();
+        expect(health, false);
```
Comment (brunoocasali) [inline test/health_test.dart:56]:
```suggestion

        expect(health, false);
```

File: test/health_test.dart [hunk 4]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
+      });
 
-    test('when the url is not valid', () async {
-      expect(client.health(), throwsException);
+      test('when the url is not valid with isHealthy', () async {
+        var health = await client.isHealthy();
+        expect(health, false);
+      });
     });
+    group("Custom dio with Http2", () {
```
Comment (brunoocasali) [inline test/health_test.dart:47]:
```suggestion

    group("Custom dio with Http2", () {
```

File: test/health_test.dart [hunk 5]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
+
+      test('when the url is not valid', () async {
+        expect(client.health(), throwsException);
+      });
 
-    test('when the url is not valid', () async {
-      expect(client.health(), throwsException);
+      test('when the url is not valid with isHealthy', () async {
+        var health = await client.isHealthy();
+        expect(health, false);
```
Comment (brunoocasali) [inline test/health_test.dart:44]:
```suggestion|

        expect(health, false);
```

File: test/health_test.dart [hunk 6]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
+
+        expect(health, {'status': 'available'});
+      });
 
-    test('of the server when the url is valid with isHealthy', () async {
-      var health = await client.isHealthy();
-      expect(health, true);
+      test('of the server when the url is valid with isHealthy', () async {
+        var health = await client.isHealthy();
+        expect(health, true);
```
Comment (brunoocasali) [inline test/health_test.dart:30]:
```suggestion

        expect(health, true);
```

File: test/health_test.dart [hunk 7]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
-      var health = await client.health();
-      expect(health, {'status': 'available'});
+      test('of the server when the url is valid', () async {
+        var health = await client.health();
+        expect(health, {'status': 'available'});
+      });
+
+      test('of the server when the url is valid with isHealthy', () async {
+        var health = await client.isHealthy();
+        expect(health, true);
```
Comment (brunoocasali) [inline test/health_test.dart:17]:
```suggestion

        expect(health, true);
```

File: test/health_test.dart [hunk 8]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
+      });
     });
+    group("Custom dio with Http2", () {
+      setUpClient(isHttp2: true);
+      test('of the server when the url is valid', () async {
```
Comment (brunoocasali) [inline test/health_test.dart:22]:
```suggestion

      test('of the server when the url is valid', () async {
```

File: test/health_test.dart [hunk 9]
```diff
@@ -4,28 +4,57 @@ import 'utils/client.dart';
...
-    setUpClient();
+    group("Default", () {
+      setUpClient();
 
-    test('of the server when the url is valid', () async {
-      var health = await client.health();
-      expect(health, {'status': 'available'});
+      test('of the server when the url is valid', () async {
+        var health = await client.health();
+        expect(health, {'status': 'available'});
```
Comment (brunoocasali) [inline test/health_test.dart:12]:
```suggestion

        expect(health, {'status': 'available'});
```

File: lib/src/http_request_impl.dart [hunk 10]
```diff
@@ -21,7 +26,15 @@ class HttpRequestImpl implements HttpRequest {
...
           ),
-        );
+        ) {
+    if (adapter != null) {
+      dio.httpClientAdapter = adapter;
+    }
+    dio.interceptors.removeImplyContentTypeInterceptor();
+    if (interceptors != null) {
+      dio.interceptors.addAll(interceptors);
+    }
```
Comment (brunoocasali) [inline lib/src/http_request_impl.dart:36]:
```suggestion
    if (adapter != null) {
      dio.httpClientAdapter = adapter;
    }
    
    dio.interceptors.removeImplyContentTypeInterceptor();
    
    if (interceptors != null) {
      dio.interceptors.addAll(interceptors);
    }
```

**Comment by brunoocasali** [comment]
> I added some tests in the health_test.dart but I have no idea how to make the server accept http2, can you help with that ?

Hi @ahmednfwela, I think we can tweak the startup of meilisearch by using this [guide](https://docs.meilisearch.com/learn/cookbooks/http2_ssl.html). But in any case I don't recommend doing it here in this SDK.

There is no advantage to doing it now since we are just exposing the ability for the end user to customize the client. This type of test should be covered on meilisearch engine side (to support http/2) and on the `dio_http2_adapter`, which is the actual implementation of http/2 handling.

File: lib/src/http_request_impl.dart (hunk 2 above)
Comment (ahmednfwela) [reply lib/src/http_request_impl.dart:9]: yes they are going to be the same, since the customDio logic is using this constructor as well

**Comment by ahmednfwela** [comment]
so I should remove the http2 tests that I added, and ignore the tests?

**Comment by brunoocasali** [comment]
> so I should remove the http2 tests that I added, and ignore the tests?

kind of, I think you can create a test that touches the new implementation `withCustomDio`.

**Comment by ahmednfwela** [comment]
I added better tests, please check them @brunoocasali

File: test/custom_dio_test.dart [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        testServer,
+        apiKey: "masterKey",
+        interceptors: [interceptor],
+      );
+
+      var health = await client.health();
+
+      expect(health, {'status': 'available'});
+      expect(interceptor.onRequestCalled, equals(true));
+    });
```
Comment (brunoocasali) [inline test/custom_dio_test.dart:43]:
```suggestion
    });

```

**Review by brunoocasali — APPROVED** [review]
Thanks again @ahmednfwela 🎉 

bors merge

**Pull request merged**